kernel/kernel_drivers.go

```diff
@@ -0,0 +1,171 @@
+		}
+		kversion = node.Name()
+	}
+	if kversion == "" {
+		return "", fmt.Errorf("no modules directory found in %q", modsDir)
+	}
+
+	return kversion, nil
+}
+
+func createFirmwareSymlinks(origin, dest string) error {
+	entries, err := os.ReadDir(origin)
+	if err != nil {
+		return err
+	}
+
+	for _, node := range entries {
+		lpath := filepath.Join(dest, node.Name())
+		origPath := filepath.Join(origin, node.Name())
+		switch node.Type() {
+		case 0, fs.ModeDir:
+			// Create link for regular files or directories
+			if err := os.Symlink(origPath, lpath); err != nil {
+				return err
+			}
+		case fs.ModeSymlink:
+			// Replicate link (it should be relative)
+			lpath := filepath.Join(dest, node.Name())
+			dest, err := os.Readlink(origPath)
+			if err != nil {
+				return err
+			}
+			if err := os.Symlink(dest, lpath); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("%q has unexpected file type: %s",
+				node.Name(), node.Type())
+		}
+	}
+
+	return nil
+}
+
+// BuildKernelDriversTree creates a tree in
+// /var/lib/snapd/kernel/<ksnapName>/<rev> with subfolders can be bind-mounted
+// on boot to /usr/lib/{modules,firmware}. This tree contains files from the
+// kernel snap mounted on kernelMount, as well as symlinks to it.
+// TODO this will be extended to consider kernel-modules components.
+func BuildKernelDriversTree(ksnapName string, rev snap.Revision, kernelMount string) (err error) {
+	// Create directory tree
+	revStr := rev.String()
+	treeRoot := filepath.Join(dirs.SnapdStateDir(dirs.GlobalRootDir), "kernel", ksnapName, revStr)
+	kversion, err := KernelVersionFromModulesDir(kernelMount)
+	if err != nil {
+		return err
+	}
+
+	// Remove tree in case something goes wrong
+	defer func() {
+		if err != nil {
+			if rmErr := os.RemoveAll(treeRoot); rmErr != nil &&
+				!errors.Is(err, fs.ErrNotExist) {
+				logger.Noticef("while cleaning up kernel tree: %v", rmErr)
+			}
+		}
+	}()
+
+	// Although empty we need "lib" because "depmod" always appends
+	// "/lib/modules/<kernel_version>" to the directory passed with option
+	// "-b".
+	modsRoot := filepath.Join(treeRoot, "lib", "modules", kversion)
+	firmwareRoot := filepath.Join(treeRoot, "lib", "firmware")
+	for _, d := range []string{modsRoot, firmwareRoot} {
+		if err := os.MkdirAll(d, 0755); err != nil {
+			return err
+		}
+	}
+
+	// Copy modinfo files from the snap
+	modsGlob := filepath.Join(kernelMount, "modules", kversion, "modules.*")
+	modFiles, err := filepath.Glob(modsGlob)
+	if err != nil {
+		// Should not really happen (only possible error is ErrBadPattern)
+		return err
+	}
+	for _, orig := range modFiles {
+		target := filepath.Join(modsRoot, filepath.Base(orig))
+		if err := osutil.CopyFile(orig, target, osutil.CopyFlagDefault); err != nil {
+			return err
+		}
+	}
+
+	// Symbolic links to early mount kernel snap
+	earlyMntDir := filepath.Join(kernelMount, "modules", kversion)
+	for _, d := range []string{"kernel", "vdso"} {
```

Comment:
IIRC vdso*.so aren't necessary for anything but debugging, and many distros don't even bother packaging them or shipping them in the default kernel package, so it's possible that we may deal with a kernel that doesn't have `vdso` directory.

Also given that those files are only for debugging, I'm not sure it's even worth bothering with them when we construct the kernel tree.